### ISSUE TITLE
support anomaly detection

### ISSUE THREAD

**Issue by int-chaos** [microsoft/FLAML#413]
Add a task type 'anomaly_detection', and at least one learner for it. 

From my understanding there are multiple types of anomalies - point anomalies, contextual anomalies, and collective anomalies - and also different types of anomaly detection algorithms - nearest-neighbor, clustering, classification, and statistic based.

If anyone has any suggestions or requests for this task, please comment.

**Comment by rashidrao-pk** [comment]
Hi @int-chaos I would like to work on this issue.

I propose to start with a minimal first PR for `task="anomaly_detection"` using `sklearn.ensemble.IsolationForest` as the first learner.

Initial scope:
- Add `anomaly_detection` as a supported task type in AutoML.
- Add an IsolationForest-based estimator.
- Support unsupervised fitting with `X_train` only.
- Support optional labeled validation data for model selection using metrics such as ROC-AUC / Average Precision when `y_val` is provided.
- Add a small example notebook and unit test.

For anomaly scores, I suggest exposing:
- `predict(X)` → normal/anomaly labels
- `decision_function(X)` or `score_samples(X)` → anomaly scores

I can first open a small design PR or draft PR so maintainers can review the API before expanding to more learners such as LocalOutlierFactor, OneClassSVM, or clustering-based methods.

**Comment by rashidrao-pk** [comment]
Hi all,

I started exploring support for anomaly detection and created a small proof-of-concept based on `sklearn.ensemble.IsolationForest`.

Current prototype:
- Added an `IsolationForestEstimator` wrapper
- Added anomaly score methods: `decision_function()` and `score_samples()`
- Added a unit test on synthetic point anomalies
- Test passes locally with:

```bash
pytest test/automl/test_anomaly_detection.py
```

Before proceeding further, I would like feedback on the desired API and integration direction.

Possible user-facing API:

```python
from flaml import AutoML

automl = AutoML()
automl.fit(
    X_train=X_train,
    task="anomaly_detection",
    time_budget=30,
)
```

A few questions:

1. Should `anomaly_detection` be added as a first-class AutoML task?
2. Should the first learner be `IsolationForest`, with later support for `OneClassSVM`, `LocalOutlierFactor`, or PyOD models?
3. Should model selection use optional labeled validation data when available, e.g. ROC-AUC / Average Precision?
4. For predictions, should FLAML expose both binary anomaly labels and continuous anomaly scores?

I am happy to continue with a small draft PR after getting feedback on the preferred design.

**Comment by thinkall** [comment]
Hi @rashidrao-pk, thanks for the great proposal — starting minimal with  IsolationForest  is exactly the right approach. Quick answers:

1. First-class task? Yes — add  `anomaly_detection`  as a proper task type (constant +  `is_anomaly_detection()`  in  `task/task.py` , register in  `GenericTask` ).
2. Learners?  `IsolationForest`  first; structure it (subclassing  `SKLearnEstimator`  in  model.py ) so  `LocalOutlierFactor` / `OneClassSVM` / `PyOD` can follow later. Keep the first PR to just IsolationForest.
3. Model selection? Use optional labels when available — default to Average Precision (or ROC-AUC) from  `score_samples` ; add the metric in  `ml.py`  and allow fit with `X_train`  only. Unsupervised proxies can come in a follow-up.
4. Predictions? Yes, expose both —  `predict` → binary labels,  `score_samples` / `decision_function`  → continuous scores (keep sklearn's sign convention).

A small draft PR to lock the API first sounds perfect — tag me for review. Thanks for picking this up!

**Comment by rashidrao-pk** [comment]
Hi @int-chaos,

Thank you for the guidance and the detailed suggestions.

Following your recommendations, I kept the first contribution minimal and focused on establishing the initial API. I have now opened a draft PR for review:

**Draft PR:** https://github.com/microsoft/FLAML/pull/1567

Current scope:
- Added `anomaly_detection` as a first-class task type.
- Added `IsolationForestEstimator` as the initial learner, subclassing `SKLearnEstimator`.
- Registered the learner for the new task.
- Exposed `predict`, `score_samples`, and `decision_function`, following scikit-learn's conventions.
- Added a unit test using a synthetic point anomaly detection dataset.

I intentionally kept the scope limited to the initial task and estimator integration, as suggested. My plan is to extend the implementation incrementally with:
- full label-free `AutoML.fit(X_train=..., task="anomaly_detection")` support,
- model selection using optional labels and anomaly-specific metrics,
- additional learners such as `LocalOutlierFactor`, `OneClassSVM`, and potentially PyOD models.

I'd really appreciate your feedback on whether this is the right direction before I continue with the next stage of the implementation.

Thank you!
